ch_tools/chadmin/internal/table.py

```diff
@@ -249,6 +261,173 @@ def delete_table(
+          AND table = '{{ table_name }}'
+    """
+
+    response = execute_query(
+        ctx,
+        query,
+        database_name=database_name,
+        table_name=table_name,
+        format_="JSON",
+    )["data"]
+
+    if 0 != len(response):
+        raise RuntimeError(
+            f"Table '{database_name}'.'{table_name}' is attached. Use delete without --detach flag."
+        )
+
+
+def check_table_engine_supported(table_metadata_path: str) -> None:
+    engine = _get_engine_table(table_metadata_path)
+    logging.info("Found engine {} in metadata {}", engine, table_metadata_path)
+
+    if "Replic" in engine:
+        raise RuntimeError(
+            f"Doesn't support removing detached table with engine {engine}"
+        )
+
+
+def _get_disk_names(ctx: Context) -> List:
+    # Disk type 'cache' of disk object_storage_cache is not supported by clickhouse-disks
+    query = """
+        SELECT name FROM system.disks
+        WHERE name!='object_storage_cache'
+    """
+    response = execute_query(
+        ctx,
+        query,
+        format_="JSON",
+    )["data"]
+
+    logging.info("Found disks: {}", response)
+    return response
+
+
+def _remove_table_data_from_disk(table_uuid: str, disk: str) -> None:
+    logging.info(
+        "_remove_table_data_from_disk: UUID={}, disk={}",
+        table_uuid,
+        disk,
+    )
+
+    object_storage_table_data_path = "store" + "/" + table_uuid[:3] + "/" + table_uuid
+
+    logging.info(
+        "Table has UUID {}, path in S3: {}.", table_uuid, object_storage_table_data_path
+    )
+
+    disk_config_path = make_ch_disks_config(disk)
+
+    code, stderr = remove_from_ch_disk(
+        disk=disk,
+        path=object_storage_table_data_path,
+        disk_config_path=disk_config_path,
+    )
+    if code:
+        raise RuntimeError(
+            f"clickhouse-disks remove command has failed: retcode {code}, stderr: {stderr.decode()}"
+        )
+
+
+def delete_detached_table(ctx, database_name, table_name):
+    logging.info("Call delete_detached_table: {}.{}", database_name, table_name)
+
+    escaped_database_name = database_name.encode("unicode_escape").decode("utf-8")
+    escaped_table_name = table_name.encode("unicode_escape").decode("utf-8")
+
+    logging.info("Escaped params: {}.{}", escaped_database_name, escaped_table_name)
+
+    check_table_dettached(ctx, escaped_database_name, escaped_table_name)
+
+    local_metadata_table_path = (
+        CLICKHOUSE_METADATA_PATH
+        + "/"
+        + escaped_database_name
+        + "/"
+        + escaped_table_name
+        + ".sql"
+    )
+
+    if not os.path.exists(local_metadata_table_path):
+        raise RuntimeError(
+            f"No metadata file for table '{escaped_database_name}'.'{escaped_table_name}' by path {local_metadata_table_path}."
+        )
+
+    check_table_engine_supported(local_metadata_table_path)
+
+    table_uuid = _get_uuid_table(local_metadata_table_path)
+    for disk in _get_disk_names(ctx):
```

Comment:
Fyi There are some bugs in `23.3-23.8` versions with loading  cache disks in `clickhouse-disks`, so I think it make sense to filter it.

Reply:
Yea, I had saw a problem with cache so I added a filter [here](https://github.com/yandex/ch-tools/pull/182/files/9b97d2fb767f051f0e72982ba28a2c975486bed1#diff-e5b2e346ca78f718d722d340262e7567c9b38626bc7b902915434df10b8f098eR344).

Reply:
Okey.
Let's do some refactoring, because it's not obvious for me that `_get_disks_names` will return all disks except 'object_storage_cache'.
 
Propose something like:
```
def _get_disk_names(ctx):
       query = "SELECT name,type FROM system.disks"

....
for disk in filter(_get_disk_names(ctx), /*remove cache disks from the list*/):
```
or 
```
def _get_disks_names(ctx, types_to_exclude):
       type_to_exclude_ch_array = /*convert types_to_exclude to string in clickhouse array format */
       query = "SELECT name,type FROM system.disks WHERE not has({type_to_exclude_ch_array}, type)"

```